scalding-core/src/main/scala/com/twitter/scalding/serialization/OrderedSerialization.scala

```diff
@@ -0,0 +1,147 @@
+
+import java.io.{ ByteArrayInputStream, InputStream, OutputStream }
+import scala.util.{ Failure, Success, Try }
+import scala.util.control.NonFatal
+
+/**
+ * In large-scale partitioning algorithms, we often use sorting.
+ * This typeclass represents something we can efficiently serialize
+ * with an added law: that we can (hopefully fast) compare the raw
+ * data.
+ */
+trait OrderedSerialization[T] extends Ordering[T] with Serialization[T] {
+  /**
+   * This compares two InputStreams. After this call, the position in
+   * the InputStreams is mutated to be the end of the record.
+   */
+  def compareBinary(a: InputStream, b: InputStream): OrderedSerialization.Result
+}
+
+object OrderedSerialization {
+  /**
+   * Represents the result of a comparison that might fail due
+   * to an error deserializing
+   */
+  sealed trait Result {
+    /**
+     * Throws if the items cannot be compared
+     */
+    def unsafeToInt: Int
+    def toTry: Try[Int]
+  }
+  /**
+   * Create a Result from an Int.
+   */
+  def resultFrom(i: Int): Result =
+    if (i > 0) Greater
+    else if (i < 0) Less
+    else Equal
+
+  def resultFrom(t: Try[Int]): Result = t match {
+    case Success(i) => resultFrom(i)
+    case Failure(e) => CompareFailure(e)
+  }
+
+  final case class CompareFailure(ex: Throwable) extends Result {
+    def unsafeToInt = throw ex
+    def toTry = Failure(ex)
+  }
+  case object Greater extends Result {
+    val unsafeToInt = 1
+    val toTry = Success(unsafeToInt)
+  }
+  case object Equal extends Result {
+    val unsafeToInt = 0
+    val toTry = Success(unsafeToInt)
+  }
+  case object Less extends Result {
+    val unsafeToInt = -1
+    val toTry = Success(unsafeToInt)
+  }
+
+  def compare[T](a: T, b: T)(implicit ord: OrderedSerialization[T]): Int =
+    ord.compare(a, b)
+
+  def compareBinary[T](a: InputStream, b: InputStream)(implicit ord: OrderedSerialization[T]): Result =
+    ord.compareBinary(a, b)
+
+  def writeThenCompare[T](a: T, b: T)(implicit ordb: OrderedSerialization[T]): Result = {
+    val abytes = Serialization.toBytes(a)
+    val bbytes = Serialization.toBytes(b)
+    val ain = new ByteArrayInputStream(abytes)
+    val bin = new ByteArrayInputStream(bbytes)
+    ordb.compareBinary(ain, bin)
+  }
+
+  /**
+   * This is slow, but always an option. Avoid this if you can, especially for large items
+   */
+  def readThenCompare[T: OrderedSerialization](as: InputStream, bs: InputStream): Result = try resultFrom {
+    val a = Serialization.read[T](as)
+    val b = Serialization.read[T](bs)
+    compare(a.get, b.get)
+  } catch {
+    case NonFatal(e) => CompareFailure(e)
+  }
+
+  /**
+   * The the serialized comparison matches the unserialized comparison
+   */
+  def compareBinaryMatchesCompare[T](implicit ordb: OrderedSerialization[T]): Law2[T] =
+    Law2("compare(a, b) == compareBinary(aBin, bBin)",
+      { (a: T, b: T) => resultFrom(ordb.compare(a, b)) == writeThenCompare(a, b) })
```

Comment:
checking `a, b` and `b, a` is probably not needed since this code is expected to be used in random generation cases (a la scalacheck). So, by doing that you are just sampling twice as many with a weird bias. If you want more test cases, just turn them up, don't put the same ones written different ways with the values reused.